Added webscraper for Thrift Savings Plan

e2thenegpii opened the pull request:
I added a webscraper to put TSP fund data into a dataframe.  In order to import pandas_datareader on my machine I had to put a requirement requests>=2.3.0

davidastephens commented:
Thanks!

Looks like the data is somewhat unreliable (see Travis Errors - I re-ran them and got the same thing).  Maybe add a check to ensure the scraper got data and a couple retries.  You can then raise RemoteDataError and add a  
```
try:
    #test code here
except RemoteDataError as e: # pragma: no cover
    raise nose.SkipTest(e)
`````
In the test.

Also, can you add a note in the what's new?

e2thenegpii commented:
I looked at why the test was failing, the CSV file that I was downloading
from TSP.gov was rather irregular regarding the placement of whitespace and
newlines.  Pandas 14.1 and older didn't handle the sloppy whitespace well
which caused the error.  I added a callback in the DataReader class that is
called when extracting the text from the response object.  This should
enable subclasses to do things like fix up sloppy responses if needed as is
the case for TSP.  The default operation is to return the response object
content.

I also added a what's new 2.3 document.

On Wed, Jan 6, 2016 at 11:37 PM, David Stephens <notifications@github.com>
wrote:

> Thanks!
>
> Looks like the data is somewhat unreliable (see Travis Errors - I re-ran
> them and got the same thing). Maybe add a check to ensure the scraper got
> data and a couple retries. You can then raise RemoteDataError and add a
>
> try:
>     #test code here
> except RemoteDataError as e: # pragma: no cover
>     raise nose.SkipTest(e)
>
> In the test.
>
> Also, can you add a note in the what's new?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pydata/pandas-datareader/pull/157#issuecomment-169547392>
> .
>

femtotrader commented on docs/source/remote_data.rst:
```diff
@@ -409,3 +410,17 @@ Company filing index from EDGAR (SEC).
+.. ipython:: python
+
+    import pandas_datareader.tsp as tsp
+    tspreader = tsp.TSPReader(start='2015-10-1',end='2015-12-31')
```
a space before `end` and after comma is missing

    tspreader = tsp.TSPReader(start='2015-10-1', end='2015-12-31')

see https://www.python.org/dev/peps/pep-0008/#whitespace-in-expressions-and-statements

femtotrader commented on pandas_datareader/base.py:
```diff
@@ -92,14 +92,22 @@ def _read_url_as_StringIO(self, url, params=None):
         return out
 
+    @classmethod
+    def _sanitize_response(cls,response):
```
```python
    def _sanitize_response(cls, response):
```

femtotrader commented on pandas_datareader/tests/test_tsp.py:
```diff
@@ -0,0 +1,16 @@
+        tspdata = tsp.TSPReader(start='2015-11-2',end='2015-11-2').read()
+        
+        assert len(tspdata == 1)
+        assert round(tspdata['I Fund'][dt.date(2015,11,2)],5) == 25.0058
```
fix space after coma

You may probably use `pandas.util.testing.assert_almost_equal`

davidastephens commented on pandas_datareader/tsp.py:
```diff
@@ -0,0 +1,63 @@
+        return {'startdate':self.start.strftime('%m/%d/%Y'),'enddate':self.end.strftime('%m/%d/%Y'),
+                'fundgroup':self.symbols,'whichButton':'CSV'}
+
+    @classmethod
```
Why is this a classmethod?

e2thenegpii replied:
The _sanitize_response method shouldn't be changing the state of the reader
object but it is providing a service for the object.  It should only be
operating on the content of the response object.

On Sun, Jan 10, 2016 at 8:30 PM, David Stephens <notifications@github.com>
wrote:

> In pandas_datareader/tsp.py
> <https://github.com/pydata/pandas-datareader/pull/157#discussion_r49285545>
> :
>
> > +    @property
> > +    def url(self):
> > +        return 'https://www.tsp.gov/InvestmentFunds/FundPerformance/index.html'
> > +
> > +    def read(self):
> > +        """ read one data from specified URL """
> > +        df = super(TSPReader, self).read()
> > +        df.columns = map(lambda x:x.strip(),df.columns)
> > +        return df
> > +
> > +    @property
> > +    def params(self):
> > +        return {'startdate':self.start.strftime('%m/%d/%Y'),'enddate':self.end.strftime('%m/%d/%Y'),
> > +                'fundgroup':self.symbols,'whichButton':'CSV'}
> > +
> > +    @classmethod
>
> Why is this a classmethod?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pydata/pandas-datareader/pull/157/files#r49285545>.
>

davidastephens replied:
Makes sense, but in that case I would use a staticmethod.

http://pythoncentral.io/difference-between-staticmethod-and-classmethod-in-python/

davidastephens commented:
Thanks - can you squash this and we can merge?

davidastephens commented:
I think this got squashed before the lint PR - can you rebase so we can merge?

Thanks.

e2thenegpii commented:
Everything should be set to go.  Thanks for your help.

davidastephens commented:
Thanks!
